Thread PG: add allreduce to threaded pg

XilunWu opened the pull request:
Summary:
Goal
Add `all_reduce` collective  to multi-threaded ProcessGroup added in D40236769 (https://github.com/pytorch/pytorch/commit/6663ae5537f3c61030ba4d425bd57a097c51430a).

Code Motion
Added `allreduce` collective to ProcessLocalGroup (a subclass of c10d ProcessGroup).

What's Next
Add a DDP test utilizing the new allreduce op.
Generalize `allreduce` to allow other `ReduceOp`s besides `SUM`.

Test Plan:
cd fbcode/caffe2
buck2 test mode/dev //caffe2/test/distributed:multi_threaded

Differential Revision: D41046606

wanchaol reviewed (changes requested):
can you rebase to master as the dependent diff landed?

XilunWu commented:
/easycla

XilunWu commented:
@wanchaol , all signals are green now.

wanchaol commented on torch/testing/_internal/distributed/multi_threaded_pg.py:
```diff
@@ -32,6 +37,21 @@ def ret_work(ret):
 
+class AllReduce:
+    def __init__(self, op):
+        assert op == ReduceOp.SUM
```
nit: throw a UnimplementedError instead of assert

wanchaol commented on torch/testing/_internal/distributed/multi_threaded_pg.py:
```diff
@@ -32,6 +37,21 @@ def ret_work(ret):
+
+    def work(self, data):
+        # data: List[List[Tensor]]
+        res = torch.empty_like(data[0][0]).copy_(data[0][0])
```
why we need to create a new tensor and copy it over? couldn't we just reuse the tensor from the first rank?

wanchaol reviewed (commented):
Can you rename the PR title and description? It looks like this PR only adds all_reduce to multithreaded pg, but it does not enable any DDP tests.

wz337 commented on test/distributed/test_multi_threaded_pg.py:
```diff
@@ -38,6 +39,21 @@ def test_broadcast_object_list(self):
+        res_num = ((0 + self.world_size - 1) * self.world_size) / 2
+        self.assertEqual(output, torch.ones(3, 3) * res_num)
+
     def test_something_else(self):
```
Noob question: do you know what is this for?

wanchaol replied:
I deleted this, looks useless lol

XilunWu replied:
@wz337 It's the guidance left by @kumpera for actual n00bs who don't know where to add more tests like me. ;-)

wz337 replied:
@XilunWu Haha thanks for providing the context. Otherwise, I would spend a long time wondering what kind of dark magic Rodrigo is trying to do here. Thanks for debunking the myth. =)

XilunWu commented:
test error exists, convert to draft.

XilunWu committed:
Thread PG: add allreduce to threaded pg (#89043)

Summary:
Goal
Add `all_reduce` collective  to multi-threaded ProcessGroup added in D40236769 (https://github.com/pytorch/pytorch/commit/6663ae5537f3c61030ba4d425bd57a097c51430a) (https://github.com/pytorch/pytorch/commit/6663ae5537f3c61030ba4d425bd57a097c51430a).

Code Motion
Added `allreduce` collective to ProcessLocalGroup (a subclass of c10d ProcessGroup).

What's Next
Add a DDP test utilizing the new allreduce op.
Generalize `allreduce` to allow other `ReduceOp`s besides `SUM`.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/89043

Test Plan:
cd fbcode/caffe2
buck2 test mode/dev //caffe2/test/distributed:multi_threaded

Differential Revision: D41046606

Pulled By: XilunWu

fbshipit-source-id: 5f41b5510ea654c5386f4d0ee0a37a387313897b

XilunWu commented:
all tests are passed now.
